### PR TITLE
Forbid usage of static inner test classes in TestNG

### PR DESCRIPTION
**Pull request by grantatspothero**
## Description

TestNG support for static inner test classes is poor

Adds a TestNG listener to fail if a test class is defined as a static inner class

## Related issues, pull requests, and links
This PR introduced usage of static inner test classes in kudu:
https://github.com/trinodb/trino/pull/10940

This PR subsequently removed the static inner test classes after some tests were not being run:
https://github.com/trinodb/trino/pull/11185

## Documentation

(X) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(X) No release notes entries required.
( ) Release notes entries required with the following suggested text:

### CODE REVIEW

File: testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportStaticInnerTestClasses.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        performActionIfInnerClass(realClass, enclosingClass ->
+            reportListenerFailure(ReportStaticInnerTestClasses.class,
+                    "Test class %s is defined as an inner class, has an enclosing class %s",
+                    realClass.getName(),
+                    enclosingClass.getName())
+        );
+    }
+
+    @VisibleForTesting
+    public static Optional<Class<?>> performActionIfInnerClass(Class<?> testClass, Consumer<Class<?>> consumer)
```
Comment (findepi) [inline testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportStaticInnerTestClasses.java:53]:
```
static boolean isInnerClass(Class<?> testClas)
```

File: testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportStaticInnerTestClasses.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.function.Consumer;
+
+import static com.google.common.base.Throwables.getStackTraceAsString;
+import static io.trino.testng.services.Listeners.reportListenerFailure;
+
+/**
+ * Detects test classes which are defined as static inner classes
+ * TestNG support is poor for these static inner test classes: https://github.com/trinodb/trino/pull/11185
+ */
+public class ReportStaticInnerTestClasses
```
Comment (findepi) [inline testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportStaticInnerTestClasses.java:17]: Remove "Static", it doesn't matter.

File: testing/trino-testing-services/src/test/java/io/trino/testng/services/TestStaticInnerTestClassVerifier.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                .contains(TestClassWithInnerTestClasses.class);
+
+        assertThat(getEnclosingClass(TestClassWithMultipleInnerTestClasses.TestInnerTestClassOne.class))
+                .isPresent()
+                .contains(TestClassWithMultipleInnerTestClasses.class);
+
+        assertThat(getEnclosingClass(TestClassWithMultipleInnerTestClasses.TestInnerTestClassTwo.class))
+                .isPresent()
+                .contains(TestClassWithMultipleInnerTestClasses.class);
+    }
```
Comment (findepi) [inline testing/trino-testing-services/src/test/java/io/trino/testng/services/TestStaticInnerTestClassVerifier.java:38]: The tested method is one-liner "is inner class". I don't think this test adds any value.

Comment (grantatspothero) [reply testing/trino-testing-services/src/test/java/io/trino/testng/services/TestStaticInnerTestClassVerifier.java:38]: I wanted to prove to myself the automation works, I can remove the `MultipleInnerTestClasses` cases but I think keeping at least the first testcase is helpful.

Comment (findepi) [reply testing/trino-testing-services/src/test/java/io/trino/testng/services/TestStaticInnerTestClassVerifier.java:38]: if/once you apply https://github.com/trinodb/trino/pull/11316/files#r819513710, there is nothing this test actually tests.

**Review by hashhar — APPROVED** [review]

**Pull request merged**